src/intercept_template.s

```diff
@@ -0,0 +1,508 @@
+ * returns. This gives the following steps:
+ *
+ * - save registers
+ * - call C function
+ * - restore registers
+ * - jump back to libc
+ *
+ * Besides this, many syscall instructions in libc are present in leaf
+ * functions. These don't necessarily have to set up the stack pointer,
+ * leaf functions can just use e.g. the address (RSP - 16) to store
+ * local variables. But they definitely can not use the memory more than
+ * 128 bytes below the stack pointer -- this area is usually referred to as
+ * red zone, see: https://en.wikipedia.org/wiki/Red_zone_(computing) .
+ * Signal handlers are examples of code that can use the stack of the current
+ * thread between any two instructions, like this code does. This leaves us
+ * with the following steps ( new steps are marked with an asterisk ) :
+ *
+ * - * decrement the stack pointer by 128
+ * - save registers
+ * - call C function
+ * - restore registers
+ * - * increment the stack pointer by 128
+ * - jump back to libc
+ *
+ * When patching libc, sometimes some instructions surrounding the original
+ * syscall instruction need to be relocated, to make space for a jump
+ * instruction that would otherwise not fit in the two byte of the syscall
+ * instruction. These instructions are moved into this assembly template.
+ * Considering these additional steps:
+ *
+ * - * execute instructions relocated from before the original syscall
+ * - decrement the stack pointer by 128
+ * - save registers
+ * - call C function
+ * - restore registers
+ * - increment the stack pointer by 128
+ * - * execute instructions relocated from after the original syscall
+ * - jump back to libc
+ *
+ * Note: the relocated instructions need to be executed while the
+ * stack pointer is not altered. This way, they can still use RSP.
+ * The only register these instructions can't rely on, is RIP. Therefore
+ * instructions such as 'mov $5, 6(%rip)', 'jmp 4', etc... can not be
+ * relocated.
+ *
+ * The arguments in the C function call ABI are passed in a way
+ * that is different from the syscall ABI. E.g. when calling the libc
+ * function called 'syscall', the first argument is the syscall number,
+ * and it is passed in RDI. A syscall instruction expects the syscall number
+ * in RAX. A conversion between the two calling conventions must be done
+ * before calling the C function.
+ *
+ * - execute instructions relocated from before the original syscall
+ * - decrement the stack pointer by 128
+ * - save registers
+ * - * rearrange arguments to use the appropriate calling convention
+ * - call C function
+ * - restore registers
+ * - increment the stack pointer by 128
+ * - execute instructions relocated from after the original syscall
+ * - jump back to libc
+ *
+ * Sometimes the C function executes the actual syscall, sometimes
+ * it calls a hook function to provide a user space implementation.
+ * In case of creating a thread ( using SYS_clone ), none of these
```

Comment:
We currently have a requirement that we support clone().  I am not sure if this is possible  based on your comments above.